cairo/ethereum/cancun/vm/precompiled_contracts/alt_bn128.cairo

```diff
@@ -171,40 +171,78 @@ func alt_bn128_mul{
     alloc_locals;
     let data = evm.value.message.value.data;
 
+    // Gas
+    let gas_cost = Uint(6000);
+    let err = charge_gas(gas_cost);
+    if (cast(err, felt) != 0) {
+        return err;
+    }
+
     tempvar u256_zero = U256(new U256Struct(0, 0));
     tempvar u256_thirty_two = U256(new U256Struct(32, 0));
     tempvar u256_sixty_four = U256(new U256Struct(64, 0));
-
     let x0_bytes = buffer_read(data, u256_zero, u256_thirty_two);
     let y0_bytes = buffer_read(data, u256_thirty_two, u256_thirty_two);
-    let n_bytes = buffer_read(data, u256_sixty_four, u256_thirty_two);
+    let k_bytes = buffer_read(data, u256_sixty_four, u256_thirty_two);
 
     let x0_value = U256_from_be_bytes(x0_bytes);
     let y0_value = U256_from_be_bytes(y0_bytes);
-    let n_value = U256_from_be_bytes(n_bytes);
+    let k_value = U256_from_be_bytes(k_bytes);
 
-    // Gas
-    let gas_cost = Uint(6000);
-    let err = charge_gas(gas_cost);
-    if (cast(err, felt) != 0) {
+    tempvar ALT_BN128_PRIME = U256(new U256Struct(alt_bn128.P_LOW_128, alt_bn128.P_HIGH_128));
+    // Check that x0 is within the interval [0, modulus)
+    let is_x0_out_of_range = U256_le(ALT_BN128_PRIME, x0_value);
+    if (is_x0_out_of_range.value != 0) {
+        tempvar err = new EthereumException(OutOfGasError);
+        return err;
+    }
+    // Check that y0 is within the interval [0, modulus)
+    let is_y0_out_of_range = U256_le(ALT_BN128_PRIME, y0_value);
+    if (is_y0_out_of_range.value != 0) {
+        tempvar err = new EthereumException(OutOfGasError);
         return err;
     }
 
-    // Operation
-    tempvar x0_value = x0_value;
-    tempvar y0_value = y0_value;
-    tempvar n_value = n_value;
-
-    tempvar data = data;
-    tempvar error: EthereumException*;
-    tempvar output: Bytes;
+    // Check that p0 is on curve
+    let x0_uint384 = uint256_to_uint384([x0_value.value]);
+    let y0_uint384 = uint256_to_uint384([y0_value.value]);
+    let p0 = G1Point(x=x0_uint384, y=y0_uint384);
+    let k_uint384 = uint256_to_uint384([k_value.value]);
```

Comment:
Put usage next to usage if possible. This k_uint384 is not related to p0 being on curve or not